File: DataStructure/Assignment/Bidding/bidding.cpp
#include <iostream>
#include <sstream>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

map<int, vector<string>, greater<int> > bids; 
string winner = "NONE";

void input() {
    int N;
    cin >> N;

    while(N--) {
        string name;
        int bid;

        cin >> name >> bid;
        bids[bid].push_back(name);
    }
}

void process() {

    for (auto& bid : bids) {
        if (bid.second.size() == 1) { //unique bid
            winner = bid.second[0];
            break;
        }
    }
}

void output() {
    cout << winner << endl;
}

int main() {

    input();

    process();

    output();

    return 0;
}

File: JavaProgramming/DataStructure/src/Bidding/Bidding.java
package Bidding;
import java.util.*;

public class Bidding {
    static Map<Integer, List<String>> bids = new TreeMap<>(Collections.reverseOrder());
    static String winner = "NONE";

    public static void main(String[] args) {
        input();
        process();
        output();
    }

    static void input() {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();

        while (N-- > 0) {
            final String name = scanner.next();
            final int bid = scanner.nextInt();
            bids.computeIfAbsent(bid, k -> new ArrayList<>()).add(name);
        }

        scanner.close();
    }

    static void process() {
        for (Map.Entry<Integer, List<String>> entry : bids.entrySet()) {
            if (entry.getValue().size() == 1) { // unique bid
                winner = entry.getValue().get(0);
                break;
            }
        }
    }

    static void output() {
        System.out.println(winner);
    }
}
